control-plane/csi-driver/src/bin/controller/controller.rs

```diff
@@ -294,185 +294,466 @@ impl VolumeTopologyMapper {
+                                return Err(Status::internal(format!(
+                                    "Error creating volume from snapshot: {}",
+                                    error
+                                )));
+                            }
+                        }
                     }
-                    None => {
-                        RestApiClient::get_client()
-                            .create_volume(
-                                &parsed_vol_uuid,
-                                replica_count,
-                                size,
-                                volume_topology,
-                                thin,
-                                sts_affinity_group_name.clone().map(AffinityGroup::new),
-                                max_snapshots,
-                                encrypted,
-                            )
-                            .await?
+                    Err(RestError::NotFound(_)) => {
+                        error!("Snapshot {} not found", snapshot_uuid);
+                        return Err(Status::not_found(format!(
+                            "Snapshot {} not found",
+                            snapshot_uuid
+                        )));
                     }
-                };
+                    Err(error) => {
+                        error!("Error checking if snapshot exists: {}", error);
+                        return Err(Status::internal(format!(
+                            "Error checking if snapshot exists: {}",
+                            error
+                        )));
+                    }
+                }
+            }
+            Some(VolumeContentSourceInfo::Volume(source_volume_uuid)) => {
+                // Create volume from another volume (clone)
+                debug!(
+                    "Creating volume clone {} from source volume {}",
+                    uuid, source_volume_uuid
+                );
 
-                // Append the 'fsId' : 'volume id' to the context if change was requested for the
-                // clone.
-                if volume.spec.content_source.is_some()
-                    && context.clone_fs_id_as_volume_id().unwrap_or(false)
-                {
-                    volume_context.insert("fsId".to_string(), volume_uuid.clone());
+                // Check if source volume exists
+                match RestApiClient::get_client().get_volume(&source_volume_uuid).await {
+                    Ok(_source_volume) => {
+                        // Create a temporary snapshot of the source volume
+                        let temp_snapshot_id = Uuid::new_v4();
+
+                        // Create the temporary snapshot
+                        let snapshot = match RestApiClient::get_client()
+                            .create_volume_snapshot(&source_volume_uuid, &temp_snapshot_id)
+                            .await 
+                        {
```

Comment:
I would suggest to keep the other code as it is and just change this section in this order:
1. Get source volume, if exists --> next step, else bail.
2. If exists create a snapshot, if fails --> delete the snapshot and bail.
3. If succeeds get the snapshot object, if exists, create restore.
4. If fails then delete the snapshot and restore, and bail.
5. If succeeds, goes through usual success flow.

Let's hear from @tiagolobocastro and @dsharma-dc